File: solve-x-main/backend/models/opportunity.go
package models

import (
	"errors"
	"time"

	"gorm.io/gorm"
)

// Opportunity represents a research/project/internship post.
type Opportunity struct {
	gorm.Model
	ProfessorID  uint       `json:"professor_id" gorm:"not null"`
	Professor    *Professor `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;foreignKey:ProfessorID;references:ID"`
	Name         string     `json:"name" gorm:"not null"`
	Details      string     `json:"details" gorm:"type:text"`
	Requirements string     `json:"requirements" gorm:"type:text"`
	Reward       string     `json:"reward" gorm:"type:text"`
	// Relationship: each opportunity can have multiple tags
	RequirementTags []Tag `json:"requirement_tags" gorm:"many2many:opportunity_tags;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`

	// Enforce type constraint
	Type string `json:"type" gorm:"type:TEXT CHECK(type IN ('research','project','internship'));not null"`
}

// CreateOpportunity creates a new opportunity
func CreateOpportunity(db *gorm.DB, professorID uint, name, details, requirements, reward, opType string, tagIDs []uint) (*Opportunity, error) {
	// Validate type
	if opType != "research" && opType != "project" && opType != "internship" {
		return nil, errors.New("invalid opportunity type")
	}

	opportunity := Opportunity{
		ProfessorID:  professorID,
		Name:         name,
		Details:      details,
		Requirements: requirements,
		Reward:       reward,
		Type:         opType,
	}

	// Create the opportunity
	if err := db.Create(&opportunity).Error; err != nil {
		return nil, err
	}

	// Associate tags if provided
	if len(tagIDs) > 0 {
		var tags []Tag
		if err := db.Find(&tags, tagIDs).Error; err != nil {
			return nil, err
		}
		if err := db.Model(&opportunity).Association("RequirementTags").Append(&tags); err != nil {
			return nil, err
		}
	}

	// Reload with associations
	if err := db.Preload("Professor").Preload("RequirementTags").First(&opportunity, opportunity.ID).Error; err != nil {
		return nil, err
	}

	return &opportunity, nil
}

func GetOpportunityByID(db *gorm.DB, id uint) (*Opportunity, error) {
	var opportunity Opportunity
	if err := db.Preload("Professor").Preload("RequirementTags").First(&opportunity, id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("opportunity not found")
		}
		return nil, err
	}
	return &opportunity, nil
}

// UpdateOpportunity updates an opportunity’s fields
func UpdateOpportunity(db *gorm.DB, id uint, updates map[string]interface{}) (*Opportunity, error) {
	opportunity, err := GetOpportunityByID(db, id)
	if err != nil {
		return nil, err
	}

	// Automatically update the "updated_at" timestamp
	updates["updated_at"] = time.Now()

	if err := db.Model(opportunity).Updates(updates).Error; err != nil {
		return nil, err
	}

	// Refresh record with relations
	if err := db.Preload("Professor").Preload("RequirementTags").First(opportunity, id).Error; err != nil {
		return nil, err
	}

	return opportunity, nil
}

// DeleteOpportunity deletes an opportunity by ID
func DeleteOpportunity(db *gorm.DB, id uint) error {
	opportunity, err := GetOpportunityByID(db, id)
	if err != nil {
		return err
	}

	if err := db.Delete(opportunity).Error; err != nil {
		return err
	}

	return nil
}

// GetOpportunitiesByProfessorID returns all opportunities created by a given professor
func GetOpportunitiesByProfessorID(db *gorm.DB, professorID uint) ([]Opportunity, error) {
	var opportunities []Opportunity
	if err := db.Preload("Professor").Preload("RequirementTags").
		Where("professor_id = ?", professorID).
		Find(&opportunities).Error; err != nil {
		return nil, err
	}

	// Return empty array instead of error when no opportunities found
	return opportunities, nil
}

// GetAllOpportunities returns all opportunities
func GetAllOpportunities(db *gorm.DB) ([]Opportunity, error) {
	var opportunities []Opportunity
	if err := db.Preload("Professor").Preload("RequirementTags").
		Order("created_at DESC").
		Find(&opportunities).Error; err != nil {
		return nil, err
	}

	return opportunities, nil
}

File: lib/services/opportunity.ts
import prisma from "@/lib/prisma";

export type OpportunityType = "research" | "project" | "internship" | "competition";

/**
 * Create a new opportunity
 */
export async function createOpportunity(
  professorId: bigint,
  name: string,
  details: string | null,
  requirements: string | null,
  reward: string | null,
  type: OpportunityType,
  tagIds?: bigint[]
): Promise<any> {
  // Validate type
  if (!["research", "project", "internship", "competition"].includes(type)) {
    throw new Error("Invalid opportunity type");
  }

  const opportunity = await prisma.opportunity.create({
    data: {
      professorId,
      name,
      details,
      requirements,
      reward,
      type,
      opportunity_tags: tagIds?.length
        ? {
            create: tagIds.map((tagId) => ({ tagId })),
          }
        : undefined,
    },
    include: {
      professor: true,
      student: true,
      opportunity_tags: {
        include: { tags: true },
      },
    },
  } as any);

  return opportunity;
}

/**
 * Get opportunity by ID
 */
export async function getOpportunityById(id: bigint): Promise<any> {
  const opportunity = await prisma.opportunity.findUnique({
    where: { id } as any,
    include: {
      professor: true,
      student: true,
      opportunity_tags: {
        include: { tags: true },
      },
    },
  } as any);

  if (!opportunity) {
    throw new Error("Opportunity not found");
  }

  return opportunity;
}

/**
 * Update opportunity fields
 */
export async function updateOpportunity(
  id: bigint,
  updates: {
    name?: string;
    details?: string;
    requirements?: string;
    reward?: string;
    type?: string;
    tagIds?: bigint[];
  }
) {
  const { tagIds, ...data } = updates;

  // If tagIds provided, delete existing and create new junction records
  if (tagIds !== undefined) {
    await prisma.opportunityTags.deleteMany({
      where: { opportunityId: id },
    });

    if (tagIds.length > 0) {
      await prisma.opportunityTags.createMany({
        data: tagIds.map((tagId) => ({ opportunityId: id, tagId })),
      });
    }
  }

  const opportunity = await prisma.opportunity.update({
    where: { id } as any,
    data,
    include: {
      professor: true,
      student: true,
      opportunity_tags: {
        include: { tags: true },
      },
    },
  } as any);

  return opportunity as any;
}

/**
 * Delete an opportunity
 */
export async function deleteOpportunity(id: bigint) {
  await prisma.opportunity.delete({ where: { id } });
}

/**
 * Get all opportunities by professor ID
 */
export async function getOpportunitiesByProfessorId(professorId: bigint): Promise<any[]> {
  return prisma.opportunity.findMany({
    where: { professorId } as any,
    include: {
      professor: true,
      student: true,
      opportunity_tags: {
        include: { tags: true },
      },
    },
    orderBy: { createdAt: "desc" },
  } as any);
}

/**
 * Get all opportunities
 */
export async function getAllOpportunities(): Promise<any[]> {
  return prisma.opportunity.findMany({
    include: {
      professor: true,
      student: true,
      opportunity_tags: {
        include: { tags: true },
      },
    },
    orderBy: { createdAt: "desc" },
  } as any);
}

/**
 * Create opportunity by owner (professor or student)
 */
export async function createOpportunityByOwner(
  ownerType: "professor" | "student",
  ownerId: bigint,
  name: string,
  details: string | null,
  requirements: string | null,
  reward: string | null,
  type: OpportunityType,
  tagIds?: bigint[]
) {
  // Basic validation
  if (!name || !type) throw new Error("Name and type are required");
  if (!ownerId) throw new Error("Owner id is required");
  if (!["research", "project", "internship", "competition"].includes(type)) {
    throw new Error("Invalid opportunity type");
  }

  // sanitize tag ids: remove falsy and convert to BigInt where needed
  const validTagIds = Array.isArray(tagIds)
    ? tagIds.filter((t) => t !== null && t !== undefined).map((t: any) => BigInt(t))
    : [];
  const data: any = {
    name,
    details,
    requirements,
    reward,
    type,
    opportunity_tags: validTagIds.length
      ? {
          create: validTagIds.map((tagId) => ({ tagId })),
        }
      : undefined,
  };

  if (ownerType === "professor") data.professorId = ownerId;
  else data.studentId = ownerId;

  try {
    const opportunity = await prisma.opportunity.create({
      data,
      include: {
        professor: true,
        student: true,
        opportunity_tags: { include: { tags: true } },
      },
    } as any);

    return opportunity as any;
  } catch (err: any) {
    console.error("prisma.opportunity.create failed", {
      data,
      error: err?.message || err,
      meta: err?.meta || null,
    });
    throw err;
  }
}

export async function getOpportunitiesByOwner(ownerType: "professor" | "student", ownerId: bigint): Promise<any[]> {
  const where: any = ownerType === "professor" ? { professorId: ownerId } : { studentId: ownerId };
  return prisma.opportunity.findMany({
    where: where as any,
    include: {
      professor: true,
      student: true,
      opportunity_tags: { include: { tags: true } },
    },
    orderBy: { createdAt: "desc" },
  } as any);
}
